test/extensions/filters/http/common/fuzz/uber_filter.cc

```diff
@@ -42,24 +42,44 @@ void UberFilterFuzzer::decode(Http::StreamDecoderFilter* filter, const test::fuz
     headers.setHost("foo.com");
   }
 
-  if (data.data().empty() && !data.has_trailers()) {
+  if (data.body_case() == test::fuzz::HttpData::BODY_NOT_SET && !data.has_trailers()) {
     end_stream = true;
   }
-  ENVOY_LOG_MISC(debug, "Decoding headers: {} ", data.headers().DebugString());
+  ENVOY_LOG_MISC(debug, "Decoding headers (end_stream={}): {} ", end_stream,
+                 data.headers().DebugString());
   const auto& headersStatus = filter->decodeHeaders(headers, end_stream);
   if (headersStatus != Http::FilterHeadersStatus::Continue &&
       headersStatus != Http::FilterHeadersStatus::StopIteration) {
     return;
   }
 
-  for (int i = 0; i < data.data().size(); i++) {
-    if (i == data.data().size() - 1 && !data.has_trailers()) {
-      end_stream = true;
+  if (data.has_http_body()) {
```

Comment:
These two blocks look very similar. Could we factor this out into something like:

```
const std::vector<std::string> data = methodThatConvertsHttpOrProtoBodyToData(data);
for (int i = 0; i < data.size(); i ++) {
   if (...) {
     end_stream = true; 
   }
   etc...
}
```

Reply:
Are we worried about performance here? Converting http data from the proto to a `std::vector` would require a full copy.

Reply:
Hmm maybe. I think `absl::StrSplit` will also copy, and then there's an unavoildable copy in creating the buffer. Maybe instead `GetSize()` and `GetSlice(int index)` can just be abstracted?

Reply:
You are right that a copy is unavoidable. I implemented the original comment